cpp/include/partition_manager.hpp

```diff
@@ -0,0 +1,171 @@
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <raft/comms/comms.hpp>
+#include <raft/handle.hpp>
+
+#include <cassert>
+#include <cmath>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <vector>
+
+namespace cugraph {
+namespace partition_2d {
+template <typename from_t>
+std::string to_string(from_t const& value)
+{
+  std::stringstream ss;
+  ss << value;
+  return ss.str();
+}
+
+// class responsible for creating 2D partition of workers:
+// responsible with finding appropriate P_ROW x P_COL
+// 2D partition and initializing the raft::handle_t communicator
+//
+// (this might be removed; or, it might exist already)
+//
+template <typename size_type = int>
+class partition_manager_t {
+ public:
+  partition_manager_t(raft::handle_t& handle, size_type p_row_size, size_type p_col_size)
+    : handle_(handle), p_row_size_(p_row_size), p_col_size_(p_col_size)
+  {
+    init_communicator();
+  }
+
+  partition_manager_t(raft::handle_t const& handle, size_type p_size) : handle_(handle)
+  {
+    partition2d(p_size);
+    init_communicator();
+  }
+
+  virtual ~partition_manager_t(void) {}
+
+ protected:
+  virtual void partition2d(size_type p_size)
+  {
+    auto sqr = static_cast<size_type>(std::sqrt(p_size));
+
+    // find divisor of p_size
+    // nearest to sqr;
+    //
+    p_row_size_ = nearest_divisor(sqr, p_size);
+    p_col_size_ = p_size / p_row_size_;
+
+    assert(p_row_size_ > 1 && p_col_size_ > 1);
+  }
+
+  virtual void init_communicator(void)
+  {
+    // TODO: init's handle's communicator (singleton?)
+  }
+
+ private:
+  raft::handle_t& handle_;
+  size_type p_row_size_;
+  size_type p_col_size_;
+
+  static decltype(auto) nearest_divisor(size_type sqr, size_type p_size)
+  {
+    assert(sqr > 0);
+
+    for (size_type div = sqr; div > 0; --div) {
+      auto p_div = p_size % div;
+      if (p_div == 0) return div;
+    }
+  }
+};
+
+// default key-naming mechanism:
+//
+struct key_naming_t {
+  key_naming_t(int row_indx,
+               int col_indx,
+               std::string const& col_suffix = std::string("_col"),
+               std::string const& row_suffix = std::string("_row"),
+               std::string const& prefix     = std::string("partition"))
+    : col_suffix_(col_suffix),
+      row_suffix_(row_suffix),
+      prefix_(prefix),
+      name_(prefix_ + "_" + to_string(row_indx) + "_" + to_string(col_indx))
```

Comment:
I think this is overly complex.

If I am not mistaken, every GPU will separate names (keys) for row-wise communication & column-wise communication, and really no need for this.

We just need one key for row sub-communiator (something as simple as "comm_p_row") and one key for column sub-communicator.

Reply:
Done.

Reply:
I think there was miscommunication,

No need for row_indx & col_indx to be added to name_.

We just need one name for row_communicator for the entire set of GPUs (e.g. comm_p_row) and one name for col_communicator for the entire set of GPUs.
```
  std::string col_name(void) const { return prefix_ + col_suffix_; 
  std::string row_name(void) const { return prefix_ + row_suffix_; }
```
should be sufficient.

Reply:
and should we really allow users to change prefix_ and (row or col)_suffix_?

To retrieve the col_key, users need to do auto col_key = key_naming_t().get_col_name(). But if handle is initialized with non-default prefix_ or (row or col)_suffix_, this will not work.

Reply:
> I think there was miscommunication,
> 
> No need for row_indx & col_indx to be added to name_.

There are two different constructors. One that takes `row_indx`, `col_indx_`; and one that does not. The semantics is that you don't _have to_ supply `row_indx`, `col_indx_`, but you can if you want (if for some reason in the future, that turns out to be useful; maybe for debug, I don't know). 
But, okay, I'll remove the constructor taking `row_indx`, `col_indx_`.

Reply:
> and should we really allow users to change prefix_ and (row or col)_suffix_?
> 
> To retrieve the col_key, users need to do auto col_key = key_naming_t().get_col_name()

I assume you'd prefer here for prefix_ and (row or col)_suffix_ to be constants and the constructor to be a default one.

Reply:
Done.